lxd/db/cluster/update.go

```diff
@@ -70,6 +70,245 @@ var updates = map[int]schema.Update{
 	31: updateFromV30,
 	32: updateFromV31,
 	33: updateFromV32,
+	34: updateFromV33,
+}
+
+// Remove multiple entries of the same volume when using remote storage.
+// Also, allow node ID to be null for the instances and storage_volumes tables, and set it to null
+// for instances and storage volumes using remote storage.
+func updateFromV33(tx *sql.Tx) error {
+	stmts := `
+SELECT storage_volumes.id, storage_volumes.name
+FROM storage_volumes
+JOIN storage_pools ON storage_pools.id=storage_volumes.storage_pool_id
+WHERE storage_pools.driver IN ("ceph", "cephfs")
+ORDER BY storage_volumes.name
+`
+
+	// Get the total number of storage volume rows.
+	count, err := query.Count(tx, "storage_volumes", "")
+	if err != nil {
+		return errors.Wrap(err, "Failed to get storage volumes count")
+	}
+
+	volumes := make([]struct {
+		ID   int
+		Name string
+	}, count)
+	dest := func(i int) []interface{} {
+		return []interface{}{
+			&volumes[i].ID,
+			&volumes[i].Name,
+		}
+	}
+
+	stmt, err := tx.Prepare(stmts)
+	if err != nil {
+		return errors.Wrap(err, "Failed to prepary storage volume query")
+	}
+
+	err = query.SelectObjects(stmt, dest)
+	if err != nil {
+		return errors.Wrap(err, "Failed to fetch storage volumes")
+	}
+
+	// Remove multiple entries of the same volume when using remote storage
+	for i := 1; i < count; i++ {
+		if volumes[i-1].Name == volumes[i].Name {
+			_, err = tx.Exec(`DELETE FROM storage_volumes WHERE id=?`, volumes[i-1].ID)
+			if err != nil {
+				return errors.Wrap(err, "Failed to delete row from storage_volumes")
+			}
+		}
+	}
+
+	stmts = `
+CREATE TABLE instances_new (
+    id INTEGER primary key AUTOINCREMENT NOT NULL,
+    node_id INTEGER,
+    name TEXT NOT NULL,
+    architecture INTEGER NOT NULL,
+    type INTEGER NOT NULL,
+    ephemeral INTEGER NOT NULL DEFAULT 0,
+    creation_date DATETIME NOT NULL DEFAULT 0,
+    stateful INTEGER NOT NULL DEFAULT 0,
+    last_use_date DATETIME,
+    description TEXT,
+    project_id INTEGER NOT NULL,
+    expiry_date DATETIME,
+    UNIQUE (project_id, name),
+    FOREIGN KEY (node_id) REFERENCES nodes (id) ON DELETE CASCADE,
+    FOREIGN KEY (project_id) REFERENCES projects (id) ON DELETE CASCADE
+);
+
+CREATE TABLE storage_volumes_new (
```

Comment:
Did you confirm that this logic doesn't break the snapshots on those custom volumes?